helpers/testsuite/test_certificates.go

```diff
@@ -0,0 +1,177 @@
+
+package testsuite
+
+import (
+	"encoding/json"
+	"os"
+	"os/exec"
+
+	"github.com/cloudflare/cfssl/csr"
+)
+
+// CreateCertificateChain creates a chain of certificates from a slice of
+// requests. The first request is the root certificate and the last is the
+// leaf. The chain is returned as a slice of PEM-encoded bytes.
+func CreateCertificateChain(requests []csr.CertificateRequest) (certChain []byte, key []byte, err error) {
+	// Create the root certificate using the first request. This will be
+	// self-signed.
+	certChain = make([]byte, 0)
+	rootCert, prevKey, err := CreateSelfSignedCert(requests[0])
+	if err != nil {
+		return nil, nil, err
+	}
+	certChain = append(certChain, rootCert...)
+
+	// For each of the next requests, create a certificate signed by the
+	// previous certificate.
+	prevCert := rootCert
+	for _, request := range requests[1:] {
+		cert, key, err := SignCertificate(request, prevCert, prevKey)
+		if err != nil {
+			return nil, nil, err
+		}
+		certChain = append(certChain, byte('\n'))
+		certChain = append(certChain, cert...)
+		prevCert = cert
+		prevKey = key
+	}
+
+	return certChain, key, nil
+}
+
+// CreateSelfSignedCert creates a self-signed certificate from a certificate
+// request. This function just calls the CLI "gencert" command.
+func CreateSelfSignedCert(request csr.CertificateRequest) (encodedCert, encodedKey []byte, err error) {
+	// Marshall the request into JSON format and write it to a temporary file.
+	jsonBytes, err := json.Marshal(request)
+	if err != nil {
+		return nil, nil, err
+	}
+	tempFile, err := createTempFile(jsonBytes)
+	if err != nil {
+		os.Remove(tempFile)
+		return nil, nil, err
+	}
+
+	// Create the certificate with the CLI tools.
+	command := exec.Command("cfssl", "gencert", "-initca", tempFile)
+	CLIOutput, err := command.CombinedOutput()
+	if err != nil {
+		os.Remove(tempFile)
+		return nil, nil, err
+	}
+	err = checkCLIOutput(CLIOutput)
+	if err != nil {
+		os.Remove(tempFile)
+		return nil, nil, err
+	}
+
+	encodedCert, err = cleanCLIOutput(CLIOutput, "cert")
+	if err != nil {
+		os.Remove(tempFile)
+		return nil, nil, err
+	}
+	encodedKey, err = cleanCLIOutput(CLIOutput, "key")
+	if err != nil {
+		os.Remove(tempFile)
+		return nil, nil, err
+	}
+
+	os.Remove(tempFile)
+
+	return encodedCert, encodedKey, nil
+}
+
+// SignCertificate uses a certificate (input as signerCert) to create a signed
+// certificate for the input request.
+func SignCertificate(request csr.CertificateRequest, signerCert, signerKey []byte) (encodedCert, encodedKey []byte, err error) {
+	// Marshall the request into JSON format and write it to a temporary file.
+	jsonBytes, err := json.Marshal(request)
+	if err != nil {
+		return nil, nil, err
+	}
+	tempJSONFile, err := createTempFile(jsonBytes)
+	if err != nil {
+		os.Remove(tempJSONFile)
+		return nil, nil, err
+	}
+
+	// Create a CSR file with the CLI tools.
+	command := exec.Command("cfssl", "genkey", tempJSONFile)
```

Comment:
similarly, we could construct a local signer from the cert and key. Generate CSR bytes and key bytes from request using csr.ParseRequest. The only tricky part is to construct the policy, which I would attach a snippet to help you:

	defaultProfile := &config.SigningProfile{
		Usage:        []string{"cert sign", "server auth"}, // so the signed certificate can be used as intermediate CA and server auth certificate. You can adjust it for your need.
		CA:           true,
		Expiry:       time.Hour,
		ExpiryString: "1h",
	}
	policy := &config.Signing{
		Profiles: map[string]*config.SigningProfile{},
		Default:  defaultProfile,
	}
	s, err := local.NewSigner(key, cert, sigAlgo, policy)